// @utam/generator/build/parser/compose-method.js
"use strict";
/*
 * Copyright (c) 2022, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildTextComposeMethod = void 0;
function toUpperCase(str) {
    return `${str.substring(0, 1).toUpperCase()}${str.substring(1)}`;
}
function buildTextComposeMethod(config, element) {
    const { addMethodDescription } = config;
    const elementName = 'name' in element ? element.name : 'root';
    const elementArgs = element.selector?.args ?? undefined;
    const name = `get${toUpperCase(elementName)}Text`;
    const method = {
        description: addMethodDescription
            ? {
                text: [`method ${name}`],
            }
            : undefined,
        name,
        args: elementArgs ? elementArgs : undefined,
        compose: [
            {
                element: elementName,
                apply: 'getText',
            },
        ],
    };
    return method;
}
exports.buildTextComposeMethod = buildTextComposeMethod;
