// @utam/compiler/build/codegen/shared.js
"use strict";
/*
 * Copyright (c) 2021, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.RETURN_SELF_ACTION = exports.UTAM_ARGUMENT_REFERENCE_ARG_TYPE = exports.UTAM_ELEMENT_REFERENCE_ARG_TYPE = exports.UTAM_LOCATOR_ARG_TYPE = exports.UTAM_ROOT_PAGE_OBJECT_ARG_TYPE = exports.UTAM_PAGE_OBJECT_ARG_TYPE = exports.UTAM_FRAME_ARG_TYPE = exports.UTAM_RETURN_CATEGORY_ROOT_PAGE_OBJECT = exports.UTAM_RETURN_CATEGORY_PAGE_OBJECT = exports.UTAM_RETURN_CATEGORY_SELF = exports.UTAM_RETURN_CATEGORY_VOID = exports.UTAM_RETURN_CATEGORY_FRAME_ELEMENT = exports.UTAM_RETURN_CATEGORY_CUSTOM_ELEMENT = exports.UTAM_RETURN_CATEGORY_BASIC_ELEMENT = exports.UTAM_RETURN_CATEGORY_PRIMITIVE = exports.UTAM_ROOT_PAGE_OBJECT_RETURN_TYPE = exports.UTAM_PAGE_OBJECT_RETURN_TYPE = exports.UTAM_FRAME_RETURN_TYPE = exports.UTAM_STRING_RETURN_TYPE = exports.UTAM_NUMBER_RETURN_TYPE = exports.UTAM_BOOLEAN_RETURN_TYPE = exports.UTAM_VOID_RETURN_TYPE = exports.UTAM_PAGE_OBJECT_CTOR = exports.UTAM_FRAME_ELEMENT_CTOR = exports.UTAM_CONTAINER_LOCATOR = exports.UTAM_CONTAINER_CTOR = exports.UTAM_ELEMENT_TYPE_TO_CTOR = exports.UTAM_ELEMENT_CTOR_BASE = exports.UTAM_ELEMENT_FACTORY_FN = exports.UTAM_ELEMENT_CTOR = exports.UTAM_ELEMENT_CTOR_NAVIGATION = exports.UTAM_ELEMENT_CTOR_DOCUMENT = exports.UTAM_FRAME_ELEMENT_TYPE = exports.UTAM_CONTAINER_TYPE = exports.UTAM_ELEMENT_DEFAULT_TYPE = exports.UTAM_BASE_ROOT_CLASS_NAME = exports.UTAM_BASE_CLASS_NAME = exports.UTAM_TYPES_LOCATOR = exports.UTAM_TYPES_ELEMENT = exports.UTAM_TYPES_DRIVER = exports.UTAM_SHADOW_CLASS = exports.UTAM_LOCATOR_BY = exports.UTAM_CORE_MODULE = exports.BEFORE_LOAD_METHOD_NAME = exports.UTAM_NAVIGATION_ELEMENT = exports.UTAM_DOCUMENT_ELEMENT = exports.UTAM_ROOT_ELEMENT = void 0;
exports.UTAM_ROOT_ELEMENT = 'root';
exports.UTAM_DOCUMENT_ELEMENT = 'document';
exports.UTAM_NAVIGATION_ELEMENT = 'navigation';
exports.BEFORE_LOAD_METHOD_NAME = '__beforeLoad__';
exports.UTAM_CORE_MODULE = '@utam/core';
exports.UTAM_LOCATOR_BY = 'By';
exports.UTAM_SHADOW_CLASS = 'ShadowRoot';
exports.UTAM_TYPES_DRIVER = 'Driver';
exports.UTAM_TYPES_ELEMENT = 'Element';
exports.UTAM_TYPES_LOCATOR = 'Locator';
// -- Page Objects Classes Names
exports.UTAM_BASE_CLASS_NAME = 'UtamBasePageObject';
exports.UTAM_BASE_ROOT_CLASS_NAME = 'UtamBaseRootPageObject';
// -- Element Types
exports.UTAM_ELEMENT_DEFAULT_TYPE = 'basic';
exports.UTAM_CONTAINER_TYPE = 'container';
exports.UTAM_FRAME_ELEMENT_TYPE = 'frame';
// -- Elements Constructors
exports.UTAM_ELEMENT_CTOR_DOCUMENT = 'DocumentUtamElement';
exports.UTAM_ELEMENT_CTOR_NAVIGATION = 'NavigationUtamElement';
exports.UTAM_ELEMENT_CTOR = 'createUtamMixinCtor';
exports.UTAM_ELEMENT_FACTORY_FN = 'createInstance';
exports.UTAM_ELEMENT_CTOR_BASE = 'BaseUtamElement';
exports.UTAM_ELEMENT_TYPE_TO_CTOR = {
    basic: 'BaseUtamElement',
    actionable: 'ActionableUtamElement',
    clickable: 'ClickableUtamElement',
    editable: 'EditableUtamElement',
    draggable: 'DraggableUtamElement',
    touchable: 'TouchableUtamElement',
};
exports.UTAM_CONTAINER_CTOR = 'ContainerCtor';
exports.UTAM_CONTAINER_LOCATOR = 'containerLocator';
exports.UTAM_FRAME_ELEMENT_CTOR = 'FrameUtamElement';
exports.UTAM_PAGE_OBJECT_CTOR = 'PageObjectCtor';
// -- Return Types
exports.UTAM_VOID_RETURN_TYPE = 'void';
exports.UTAM_BOOLEAN_RETURN_TYPE = 'boolean';
exports.UTAM_NUMBER_RETURN_TYPE = 'number';
exports.UTAM_STRING_RETURN_TYPE = 'string';
exports.UTAM_FRAME_RETURN_TYPE = 'frame';
exports.UTAM_PAGE_OBJECT_RETURN_TYPE = 'pageObject';
exports.UTAM_ROOT_PAGE_OBJECT_RETURN_TYPE = 'rootPageObject';
// -- Return Type Category
exports.UTAM_RETURN_CATEGORY_PRIMITIVE = 'primitive';
exports.UTAM_RETURN_CATEGORY_BASIC_ELEMENT = 'basicElement';
exports.UTAM_RETURN_CATEGORY_CUSTOM_ELEMENT = 'customElement';
exports.UTAM_RETURN_CATEGORY_FRAME_ELEMENT = 'frameElement';
exports.UTAM_RETURN_CATEGORY_VOID = 'void';
exports.UTAM_RETURN_CATEGORY_SELF = 'self';
exports.UTAM_RETURN_CATEGORY_PAGE_OBJECT = 'pageObject';
exports.UTAM_RETURN_CATEGORY_ROOT_PAGE_OBJECT = 'rootPageObject';
// -- Arg Types
exports.UTAM_FRAME_ARG_TYPE = 'frame';
exports.UTAM_PAGE_OBJECT_ARG_TYPE = 'pageObject';
exports.UTAM_ROOT_PAGE_OBJECT_ARG_TYPE = 'rootPageObject';
exports.UTAM_LOCATOR_ARG_TYPE = 'locator';
exports.UTAM_ELEMENT_REFERENCE_ARG_TYPE = 'elementReference';
exports.UTAM_ARGUMENT_REFERENCE_ARG_TYPE = 'argumentReference';
// -- Hardcoded action
exports.RETURN_SELF_ACTION = 'returnSelf';
