// @utam/core/build/index.js
"use strict";
/*
 * Copyright (c) 2021, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_POLLING_INTERVAL = exports.DEFAULT_EXPLICIT_WAIT_TEIMEOUT = exports.isElementAttachedToDom = exports.NavigationUtamElement = exports.DocumentUtamElement = exports.UtamPageObjectProvider = exports.createInstanceAndWaitForRoot = exports.createInstance = exports.wait = exports.checkLocator = exports.By = exports.UtamBaseRootPageObject = exports.UtamBasePageObject = exports.ShadowRoot = void 0;
/* eslint-disable @typescript-eslint/no-unused-vars */
const document_utam_element_1 = require("./document-utam-element");
Object.defineProperty(exports, "DocumentUtamElement", { enumerable: true, get: function () { return document_utam_element_1.DocumentUtamElement; } });
const navigation_utam_element_1 = require("./navigation-utam-element");
Object.defineProperty(exports, "NavigationUtamElement", { enumerable: true, get: function () { return navigation_utam_element_1.NavigationUtamElement; } });
const helper_methods_1 = require("./helper-methods");
Object.defineProperty(exports, "createInstance", { enumerable: true, get: function () { return helper_methods_1.createInstance; } });
Object.defineProperty(exports, "createInstanceAndWaitForRoot", { enumerable: true, get: function () { return helper_methods_1.createInstanceAndWaitForRoot; } });
Object.defineProperty(exports, "wait", { enumerable: true, get: function () { return helper_methods_1.wait; } });
const by_1 = require("./lib/by");
Object.defineProperty(exports, "By", { enumerable: true, get: function () { return by_1.By; } });
Object.defineProperty(exports, "checkLocator", { enumerable: true, get: function () { return by_1.checkLocator; } });
const shadow_element_1 = require("./shadow-element");
Object.defineProperty(exports, "ShadowRoot", { enumerable: true, get: function () { return shadow_element_1.ShadowRoot; } });
const dom_traversal_1 = require("./helpers/dom-traversal");
Object.defineProperty(exports, "isElementAttachedToDom", { enumerable: true, get: function () { return dom_traversal_1.isElementAttachedToDom; } });
const utam_page_object_1 = require("./utam-page-object");
Object.defineProperty(exports, "UtamBasePageObject", { enumerable: true, get: function () { return utam_page_object_1.UtamBasePageObject; } });
Object.defineProperty(exports, "UtamBaseRootPageObject", { enumerable: true, get: function () { return utam_page_object_1.UtamBaseRootPageObject; } });
const utam_page_object_provider_1 = require("./utam-page-object-provider");
Object.defineProperty(exports, "UtamPageObjectProvider", { enumerable: true, get: function () { return utam_page_object_provider_1.UtamPageObjectProvider; } });
__exportStar(require("./base-utam-element"), exports);
__exportStar(require("./constants"), exports);
__exportStar(require("./frame-utam-element"), exports);
__exportStar(require("./helpers/create-utam-element"), exports);
__exportStar(require("./primitive-utam-elements"), exports);
exports.DEFAULT_EXPLICIT_WAIT_TEIMEOUT = 20000;
exports.DEFAULT_POLLING_INTERVAL = 200;
