// @utam/core/build/utam-page-object-provider.js
"use strict";
/*
 * Copyright (c) 2021, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.UtamPageObjectProvider = void 0;
/**
 * Class that exposes static adapters between the UTAM Loader and the UTAM Core runtime.
 * This class is responsible for passing data from the loader to the runtime.
 * The adapter functions are bound to specific loader's APIs in its constructor.
 */
class UtamPageObjectProvider {
    /**
     * Return a page object constructor that represents the concrete implementation of the page object interface
     * passed as an argument.
     *
     * @param typeCtor page object that represent an interface
     * @returns a page object that is the concrete implementation of the given interface
     */
    static getPageObjectImpl(typeCtor) {
        if (!UtamPageObjectProvider.pageObjectImplAdapter) {
            throw new Error(`Error: Loader cannot resolve any PageObject: Implementation Adapter not provided`);
        }
        return UtamPageObjectProvider.pageObjectImplAdapter(typeCtor);
    }
    /**
     * Set the function used to find a page object implementation that matches an interface for a given active
     * profile. This function is bound to the loader function that finds an implementation matching an interface in
     * the loader constructor.
     *
     * @param implAdapter adapter function used to find page object implementations
     */
    static setPageObjectImplAdapter(implAdapter) {
        UtamPageObjectProvider.pageObjectImplAdapter = implAdapter;
    }
    /**
     * Set the adapter function for getting the bridgeAppTitle
     * @param bridgeAppTitleAdapter adapter function used to find the bridgeAppTitle set in the loader config
     */
    static setGetBridgeAppTitleAdapter(bridgeAppTitleAdapter) {
        UtamPageObjectProvider.bridgeAppTitleAdapter = bridgeAppTitleAdapter;
    }
    /**
     * Return the bridgeAppTitle if an adapter has been set (loader has been instanciated), undefined otherwise
     */
    static getBridgeAppTitle() {
        return UtamPageObjectProvider.bridgeAppTitleAdapter
            ? UtamPageObjectProvider.bridgeAppTitleAdapter()
            : undefined;
    }
}
exports.UtamPageObjectProvider = UtamPageObjectProvider;
